core/src/main/java/org/apache/iceberg/DataFiles.java

```diff
@@ -259,11 +268,19 @@ public Builder withFileSizeInBytes(long newFileSizeInBytes) {
     }
 
     public Builder withPartitionPath(String newPartitionPath) {
+      return withPartitionPath(newPartitionPath, DEFAULT_PARTITION_STRING_CONVERSION);
+    }
+
+    public Builder withPartitionPath(
+        String newPartitionPath, BiFunction<String, Type, Object> partitionPathConverter) {
```

Comment:
If there's general support for this idea, I'll add a unit test that exercises a simple custom conversion logic.

Reply:
Any system's we are thinking about here? I know we are pretty much good for Hive now, do we have custom apps we are trying to support?

Reply:
Yes, let me update the description to be clear. But In short, one use case is for Delta Uniform Iceberg metadata conversion. Currently, timestamp types are not supported and Iceberg's `Conversion` utility https://github.com/apache/iceberg/blob/main/api/src/main/java/org/apache/iceberg/types/Conversions.java#L69 does not handle timestamps.

I was a bit hestitant to add another case for Timestamp to that utility since Iceberg does not specify (and arguably *should not*) specify how to convert string partition values into a given type. For now it's all Hive style assumptions which was done to facilitate migration use cases. 

Since Iceberg does not specify this conversion behavior exposing an API which allows people to pass in their own converter was the first approach that came to mind for solving the problem I had but arguably if that's the only use case, one could also make the case that Delta Uniform should take the responsibility of converting to Partition tuples and using the existing DataFile APIs? Def open to that as well.

Reply:
Possible Alternatives:

1. No new APIs, callers (in this case Uniform) should take the responsibiity of computing partition tuples themselves and use the `withPartition(StructLike partition)` API. 
2. Add the conversion logic for timestamps to Conversion. I wanted to avoid this mostly since Iceberg doesn't specify how to convert strings to partition values so whatever would be implemented would have some set of assumptions (iso8601 string with some assumed precision etc), and I didn't want to set a precedent that there is some established conversion behavior that should be followed (since there isn't). But there's a counter argument to that which is, really the only remaining type that's missing in that logic is Timestamps anyways.

Reply:
I like both of those alternatives. 

1. Is probably the right solution (our string support is there pretty much expressly for hive) but I have no problem with doing 2. since we don't really have any other use case for this.